Multiple player dev

koolskateguy89 opened the pull request:
Start of attempt to solve #16 (request to support other players)

- Added checking which player is open ([linear time ;)](https://github.com/mheine/gnome-shell-spotify-label/issues/16#issuecomment-781737066) )
- 'Verified' that VLC & Elisa work - if the media file has defined Title & Artist tags

Also [extension.js:59](https://github.com/mheine/gnome-shell-spotify-label/blob/11e3563168b7cfe5b45e08bae53222945d0dc683/extension.js#L59) there should be `let/var` before `children` as it's previously undefined and was giving JS Warnings in Gnome-SHELL logs.

I still need to check the other players and check which other players support MPRIS, including Clementine & RhythmBox that the issue OP wanted (I mean add it to the `players` object and check it works).

koolskateguy89 committed:
Start testing multiple players & test vlc

koolskateguy89 committed:
Fix checking which window is open & check Elisa

koolskateguy89 commented:
I have no idea how to request review....

koolskateguy89 commented:
So [apparently the PR author can't request review](https://stackoverflow.com/a/41028889) so I'm just have to tag you @mheine  :)

koolskateguy89 commented:
Note: I just got a `JS ERROR: get_meta_window(...).get_wm_class(...) is null` - no idea why.
I'll try and reproduce it tomorrow despite having no idea what caused it 😬

mheine commented:
Assigned myself as a reviewer, hoping to add some thoughts and input when I find the time :technologist:

koolskateguy89 committed:
Test Rhythmbox & JuK

koolskateguy89 commented:
Thread about checking which player is open

https://unix.stackexchange.com/a/480820

koolskateguy89 commented:
- Gwenview is an image viewer (seems to also be a video viewer but I couldn't get that to work and gave up after like 5 seconds) so I removed it
- Okular is a [PDF viewer](https://okular.kde.org/) so I removed it

- Amarok seems to no longer be maintained and I can't find a way to install it without compiling the source code so I can't confirm it works
- I couldn't install [Vvave](https://github.com/KDE/vvave) when following the steps in their README (doing `make` gave error message `No targets specified and no makefile found. Stop.`) so I can't confirm it works
- Clementine works fine - again as long as the media has defined Artist & Title tags

- Plasma media center [requires KDE Plasma to be installed](https://github.com/KDE/plasma-mediacenter)
- So does [Plasma browser integration](https://github.com/KDE/plasma-browser-integration)

(I'm on Ubuntu with Gnome Shell and not willing to install KDE Plasma tbh)

Every other player works for me and issue #16 (specifically Rhythmbox & Clementine) is solved :)

@mheine

koolskateguy89 commented:
However checking the open player is difficult, the thread I posted earlier uses `qdbus` which when I run, gives an error `could not find a Qt installation of ''` so ¯\\\_(ツ)\_/¯

koolskateguy89 committed:
Finished I think

koolskateguy89 commented:
School starts again for me tomorrow so unfortunately I'll be quite inactive o7

koolskateguy89 committed:
Use ByteArray.toString()

Protects against future behaviour change

mavit commented on extension.js:
```diff
@@ -272,6 +289,38 @@ function toggleWindow() {
+		if (players[windowName])
+			return players[windowName];
+	}
+}
```
What about players that don't have windows?  E.g., https://github.com/mavit/slimpris2/

It make more sense to ask DBus for a list of available players.  E.g.,

```bash
dbus-send --print-reply --dest=org.freedesktop.DBus /org/freedesktop/DBus org.freedesktop.DBus.ListNames | grep -E '^ +string "org\.mpris\.MediaPlayer2\.'
```

mavit reviewed (commented):
A couple of general suggestions:

- It might turn out to be easier to [access DBus directly](https://wiki.gnome.org/Gjs/Examples/DBusClient) rather than shelling out to `dbus-send`.
- Gnome Shell already features [an MPRIS client that is used to populate the now-playing info in the notification area](https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/master/js/ui/mpris.js).  Perhaps this code could be substantially reused?  

Hope this helps!

koolskateguy89 committed:
Merge branch 'mheine:master' into multiple-player-dev
